src/jitlayers.h

```diff
@@ -30,23 +27,58 @@ extern "C" {
 extern TargetMachine *jl_TargetMachine;
 extern Module *shadow_output;
 extern bool imaging_mode;
-#if defined(_OS_WINDOWS_) && defined(_CPU_X86_64_)
-extern Function *juliapersonality_func;
-#endif
-
-
-typedef struct {Value *gv; int32_t index;} jl_value_llvm; // uses 1-based indexing
 
 void addTargetPasses(legacy::PassManagerBase *PM, TargetMachine *TM);
 void addOptimizationPasses(legacy::PassManagerBase *PM, int opt_level, bool dump_native=false);
-void* jl_emit_and_add_to_shadow(GlobalVariable *gv, void *gvarinit = NULL);
-void* jl_get_globalvar(GlobalVariable *gv);
-GlobalVariable *jl_get_global_for(const char *cname, void *addr, Module *M);
-void jl_add_to_shadow(Module *m);
-void jl_init_function(Function *f);
-bool jl_can_finalize_function(StringRef F);
-void jl_finalize_function(StringRef F);
-void jl_finalize_module(Module *m, bool shadow);
+void jl_finalize_module(std::unique_ptr<Module>  m);
+void jl_merge_module(Module *dest, std::unique_ptr<Module> src);
+
+typedef struct _jl_llvm_functions_t {
+    std::string functionObject;     // jlcall llvm Function name
+    std::string specFunctionObject; // specialized llvm Function name
+} jl_llvm_functions_t;
+
+struct jl_returninfo_t {
+    llvm::Function *decl;
+    enum CallingConv {
+        Boxed = 0,
+        Register,
+        SRet,
+        Union,
+        Ghosts
+    } cc;
+    size_t union_bytes;
+    size_t union_align;
+    size_t union_minalign;
+};
+
+typedef std::vector<std::tuple<jl_method_instance_t*, jl_returninfo_t::CallingConv, llvm::Function*, jl_value_t*, bool>> jl_codegen_call_targets_t;
+typedef std::tuple<std::unique_ptr<Module>, jl_llvm_functions_t, jl_value_t*> jl_compile_result_t;
+
+typedef struct {
+    // outputs
+    jl_codegen_call_targets_t workqueue;
+    std::map<void *, GlobalVariable *> globals;
+    std::map<jl_datatype_t *, DIType *> ditypes;
+    std::map<jl_datatype_t *, Type *> llvmtypes;
+    // inputs
+    size_t world = 0;
+    const jl_cgparams_t *params = &jl_default_cgparams;
+    bool cache = false;
+} jl_codegen_params_t;
```

Comment:
do you intend for this to be reusable across code generators, or just sample usage like jl_create_native?

Reply:
It can be reusable. It's mostly intended to represent a single `llvm::Module` of output, but we can reset it when you compile the result (empty the `globals` list and reset the module handle) to prepare it for more reuse.